# ***NYC Taxi Date Analysis***

* **Author Name:** Tuhid Khan
* **Email:** tuhidkhan1202@gmail.com

# About Dataset
Here You can find the data on Kaggle [link](https://www.kaggle.com/datasets/yasserh/nyc-taxi-trip-duration)

# NYC Taxi Trip Duration Dataset

## About the Dataset
This dataset is based on the 2016 NYC Yellow Cab trip record data published by the New York City Taxi and Limousine Commission (TLC). It contains key details about individual taxi rides in New York City, including pickup and dropoff times, coordinates, passenger counts, and overall trip durations. 

Primary objectives typically include:
- Performing Exploratory Data Analysis (EDA) on urban mobility patterns.
- Predicting the total duration of a ride (`trip_duration`) based on spatial and temporal features.

---

## Dataset Overview & Metadata

- **Source / Owner:** NYC Taxi & Limousine Commission (TLC) via Kaggle (`yasserh/nyc-taxi-trip-duration`)
- **Primary File:** `nyc_taxi_trip_duration.csv` (or `train.csv`)
- **License:** CC0: Public Domain
- **Dataset Shape:** ~729,322 Rows × 11 Columns

---

## Data Schema & Variable Description

| Column Name | Data Type | Description |
| :--- | :--- | :--- |
| `id` | String / Object | Unique identifier for each trip |
| `vendor_id` | Integer / Categorical | Code indicating the provider associated with the trip record (e.g., 1 or 2) |
| `pickup_datetime` | Datetime | Date and time when the meter was engaged |
| `dropoff_datetime` | Datetime | Date and time when the meter was disengaged |
| `passenger_count` | Integer | The number of passengers in the vehicle (driver-entered value) |
| `pickup_longitude` | Float | Longitude coordinate where the meter was engaged |
| `pickup_latitude` | Float | Latitude coordinate where the meter was engaged |
| `dropoff_longitude`| Float | Longitude coordinate where the meter was disengaged |
| `dropoff_latitude` | Float | Latitude coordinate where the meter was disengaged |
| `store_and_fwd_flag`| Categorical (Y/N)| Indicates if trip record was held in vehicle memory due to server connection loss (`Y` = Store & forward, `N` = No) |
| `trip_duration` | Integer | **Target Variable:** Total duration of the trip in seconds |

---

## Variable Classification

- **Continuous Numerical:** `pickup_longitude`, `pickup_latitude`, `dropoff_longitude`, `dropoff_latitude`, `trip_duration`
- **Discrete Numerical:** `vendor_id`, `passenger_count`
- **Temporal / Datetime:** `pickup_datetime`, `dropoff_datetime`
- **Nominal Categorical:** `store_and_fwd_flag`

# 1. DATA LOADING
 - Import pandas 
 - Load dataset using pd.read_csv() 
 - Display first 5 rows 
 - Check dataset shape and columns 

In [1]:
# importing the required libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as ps

In [2]:
# Load dataset using pd.read_csv()
df = pd.read_csv("nyc_taxi_trip.csv")


In [3]:
# Display first 5 rows 
df.head()

,id,vendor_id,pickup_datetime,dropoff_datetime,passenger_count,pickup_longitude,pickup_latitude,dropoff_longitude,dropoff_latitude,store_and_fwd_flag,trip_duration
0,id2875421,2,2016-03-14 17:24:55,2016-03-14 17:32:30,1,-73.982155,40.767937,-73.964630,40.765602,N,455
1,id2377394,1,2016-06-12 00:43:35,2016-06-12 00:54:38,1,-73.980415,40.738564,-73.999481,40.731152,N,663
2,id3858529,2,2016-01-19 11:35:24,2016-01-19 12:10:48,1,-73.979027,40.763939,-74.005333,40.710087,N,2124
3,id3504673,2,2016-04-06 19:32:31,2016-04-06 19:39:40,1,-74.010040,40.719971,-74.012268,40.706718,N,429
4,id2181028,2,2016-03-26 13:30:55,2016-03-26 13:38:10,1,-73.973053,40.793209,-73.972923,40.782520,N,435


In [4]:
# Check dataset columns 
df.columns

Index(['id', 'vendor_id', 'pickup_datetime', 'dropoff_datetime',
       'passenger_count', 'pickup_longitude', 'pickup_latitude',
       'dropoff_longitude', 'dropoff_latitude', 'store_and_fwd_flag',
       'trip_duration'],
      dtype='str')

In [5]:
# Check dataset shape
df.shape

(1458644, 11)

In [6]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1458644 entries, 0 to 1458643
Data columns (total 11 columns):
 #   Column              Non-Null Count    Dtype  
---  ------              --------------    -----  
 0   id                  1458644 non-null  str    
 1   vendor_id           1458644 non-null  int64  
 2   pickup_datetime     1458644 non-null  str    
 3   dropoff_datetime    1458644 non-null  str    
 4   passenger_count     1458644 non-null  int64  
 5   pickup_longitude    1458644 non-null  float64
 6   pickup_latitude     1458644 non-null  float64
 7   dropoff_longitude   1458644 non-null  float64
 8   dropoff_latitude    1458644 non-null  float64
 9   store_and_fwd_flag  1458644 non-null  str    
 10  trip_duration       1458644 non-null  int64  
dtypes: float64(4), int64(3), str(4)
memory usage: 122.4 MB


# DATA CLEANING 
- Check missing values 
- Handle missing values 
- Remove duplicates 
- Convert datetime columns 
- Remove invalid rows (e.g. zero distance, zero duration)

In [7]:
df.isnull().sum()

id                    0
vendor_id             0
pickup_datetime       0
dropoff_datetime      0
passenger_count       0
pickup_longitude      0
pickup_latitude       0
dropoff_longitude     0
dropoff_latitude      0
store_and_fwd_flag    0
trip_duration         0
dtype: int64

- No missing values

In [8]:

df.duplicated().sum()

np.int64(0)

- No dublicate rows

In [9]:
# Convert datetime columns
df['pickup_datetime'] = pd.to_datetime(df['pickup_datetime'])
df['dropoff_datetime'] = pd.to_datetime(df['dropoff_datetime'])

In [10]:
df.head()

,id,vendor_id,pickup_datetime,dropoff_datetime,passenger_count,pickup_longitude,pickup_latitude,dropoff_longitude,dropoff_latitude,store_and_fwd_flag,trip_duration
0,id2875421,2,2016-03-14 17:24:55,2016-03-14 17:32:30,1,-73.982155,40.767937,-73.964630,40.765602,N,455
1,id2377394,1,2016-06-12 00:43:35,2016-06-12 00:54:38,1,-73.980415,40.738564,-73.999481,40.731152,N,663
2,id3858529,2,2016-01-19 11:35:24,2016-01-19 12:10:48,1,-73.979027,40.763939,-74.005333,40.710087,N,2124
3,id3504673,2,2016-04-06 19:32:31,2016-04-06 19:39:40,1,-74.010040,40.719971,-74.012268,40.706718,N,429
4,id2181028,2,2016-03-26 13:30:55,2016-03-26 13:38:10,1,-73.973053,40.793209,-73.972923,40.782520,N,435


In [11]:
# Remove invalid rows (e.g. zero distance, zero duration)
df = df[
    (df['trip_duration'] > 0) & 
    (df['passenger_count'] > 0) &
    (df['pickup_latitude'].between(40.5, 41.0)) &
    (df['pickup_longitude'].between(-74.3, -73.7)) &
    (df['dropoff_latitude'].between(40.5, 41.0)) &
    (df['dropoff_longitude'].between(-74.3, -73.7))
]


df.head()

,id,vendor_id,pickup_datetime,dropoff_datetime,passenger_count,pickup_longitude,pickup_latitude,dropoff_longitude,dropoff_latitude,store_and_fwd_flag,trip_duration
0,id2875421,2,2016-03-14 17:24:55,2016-03-14 17:32:30,1,-73.982155,40.767937,-73.964630,40.765602,N,455
1,id2377394,1,2016-06-12 00:43:35,2016-06-12 00:54:38,1,-73.980415,40.738564,-73.999481,40.731152,N,663
2,id3858529,2,2016-01-19 11:35:24,2016-01-19 12:10:48,1,-73.979027,40.763939,-74.005333,40.710087,N,2124
3,id3504673,2,2016-04-06 19:32:31,2016-04-06 19:39:40,1,-74.010040,40.719971,-74.012268,40.706718,N,429
4,id2181028,2,2016-03-26 13:30:55,2016-03-26 13:38:10,1,-73.973053,40.793209,-73.972923,40.782520,N,435




 filters the dataset to remove invalid or bad data. 

* **`trip_duration > 0`**: Keeps trips with a duration greater than zero seconds.
* **`passenger_count > 0`**: Ensures there is at least one passenger.
* **`pickup/dropoff latitude & longitude`**: Limits locations to the New York area (latitudes between 40.5 and 41.0, longitudes between -74.3 and -73.7) to remove error coordinates like `0, 0`.
  


---
# FEATURE ENGINEERING 
- Extract hour, day, and month from datetime 
- Calculate trip duration (if needed) 
- Calculate distance using coordinates (Haversine formula) 
- Create new useful features 

In [12]:
# Convert datetime columns 
df['hour'] = df['pickup_datetime'].dt.hour
df['day_of_week'] = df['pickup_datetime'].dt.day_name()
df['month'] = df['pickup_datetime'].dt.month



* **`hour`**: Extracts the hour (0–23).
* **`day_of_week`**: Extracts the day name (e.g., Monday).
* **`month`**: Extracts the month number (1–12).

In [13]:
#Calculate trip duration
df['trip_duration'] = df['dropoff_datetime'] - df['pickup_datetime']
df['trip_duration'] = df['trip_duration'].dt.total_seconds() / 60


df.trip_duration.mean()


np.float64(15.975417013701653)

### 15.9 min
We calculate the average trip duration in minutes to understand typical travel times and spot unusually long or short trips in the dataset.

In [14]:

def haversine_distance(lat1, lon1, lat2, lon2):
    R = 6371  # Earth radius in kilometers

    lat1 = np.radians(lat1)
    lon1 = np.radians(lon1)
    lat2 = np.radians(lat2)
    lon2 = np.radians(lon2)

    dlat = lat2 - lat1
    dlon = lon2 - lon1

    a = np.sin(dlat / 2) ** 2 + np.cos(lat1) * np.cos(lat2) * np.sin(dlon / 2) ** 2
    c = 2 * np.arcsin(np.sqrt(a))

    return R * c

df['trip_distance'] = haversine_distance(
    df['pickup_latitude'],
    df['pickup_longitude'],
    df['dropoff_latitude'],
    df['dropoff_longitude']
)
df.head()

,id,vendor_id,pickup_datetime,dropoff_datetime,passenger_count,pickup_longitude,pickup_latitude,dropoff_longitude,dropoff_latitude,store_and_fwd_flag,trip_duration,hour,day_of_week,month,trip_distance
0,id2875421,2,2016-03-14 17:24:55,2016-03-14 17:32:30,1,-73.982155,40.767937,-73.964630,40.765602,N,7.583333,17,Monday,3,1.498521
1,id2377394,1,2016-06-12 00:43:35,2016-06-12 00:54:38,1,-73.980415,40.738564,-73.999481,40.731152,N,11.050000,0,Sunday,6,1.805507
2,id3858529,2,2016-01-19 11:35:24,2016-01-19 12:10:48,1,-73.979027,40.763939,-74.005333,40.710087,N,35.400000,11,Tuesday,1,6.385098
3,id3504673,2,2016-04-06 19:32:31,2016-04-06 19:39:40,1,-74.010040,40.719971,-74.012268,40.706718,N,7.150000,19,Wednesday,4,1.485498
4,id2181028,2,2016-03-26 13:30:55,2016-03-26 13:38:10,1,-73.973053,40.793209,-73.972923,40.782520,N,7.250000,13,Saturday,3,1.188588



---
# FEATURE SCALING / NORMALIZATION 
- Apply Min-Max scaling 
- Apply Standardization 
- Compare both approaches

In [15]:

from sklearn.preprocessing import MinMaxScaler, StandardScaler

cols_to_scale = ["trip_duration", "trip_distance", "passenger_count", "hour"]

# Min-Max Scaling
minmax_scaler = MinMaxScaler()
df_minmax = df.copy()
df_minmax[cols_to_scale] = minmax_scaler.fit_transform(df[cols_to_scale])

# Standardization
standard_scaler = StandardScaler()
df_standard = df.copy()
df_standard[cols_to_scale] = standard_scaler.fit_transform(df[cols_to_scale])

# Compare
comparison = pd.DataFrame({
    "Original_mean": df[cols_to_scale].mean(),
    "Original_std": df[cols_to_scale].std(),
    "MinMax_mean": df_minmax[cols_to_scale].mean(),
    "MinMax_std": df_minmax[cols_to_scale].std(),
    "Standard_mean": df_standard[cols_to_scale].mean(),
    "Standard_std": df_standard[cols_to_scale].std(),
})

comparison

,Original_mean,Original_std,MinMax_mean,MinMax_std,Standard_mean,Standard_std
trip_duration,15.975417,87.292573,0.000272,0.001485,-3.615145e-17,1.0
trip_distance,3.423184,3.884150,0.071667,0.081318,-6.520521e-17,1.0
passenger_count,1.664722,1.314375,0.083090,0.164297,-1.013177e-16,1.0
hour,13.606781,6.398876,0.591599,0.278212,-7.409683e-18,1.0


### Comparison: Min-Max vs Standardization

- **Min-Max scaling** squeezed all values between 0 and 1,  but the mean/std still differ per column depending on outliers (e.g. `trip_duration` mean is almost 0 due to extreme long trips).
  
- **Standardization** gave every column a mean ≈ 0 and std = 1, making it more consistent and reliable — this makes it the better choice for this dataset.


---
# 5. FEATURE ENCODING 
- Encode categorical features (e.g. passenger count categories) 
- Use label encoding and one-hot encoding 

In [16]:
df['passenger_count'].value_counts()

passenger_count
1    1032796
2     210165
5      78060
3      59862
6      48318
4      28381
7          2
9          1
8          1
Name: count, dtype: int64

In [17]:

from sklearn.preprocessing import LabelEncoder

# 1. Map numerical counts into categorical bins
def categorize_passengers(count):
    if count == 1:
        return 'Single'
    elif count <= 4:
        return 'Small Group'
    else:
        return 'Large Group'

df['passenger_category'] = df['passenger_count'].apply(categorize_passengers)

# 2. Apply Label Encoding
le = LabelEncoder()
df['passenger_category_label'] = le.fit_transform(df['passenger_category'])

# Preview results
df[['passenger_count', 'passenger_category', 'passenger_category_label']].head()

,passenger_count,passenger_category,passenger_category_label
0,1,Single,1
1,1,Single,1
2,1,Single,1
3,1,Single,1
4,1,Single,1


### Step 1: Feature Binning & Label Encoding

* **`categorize_passengers()` Function**: Groups the raw continuous numerical passenger counts into three distinct categories:
  * **`Single`**: 1 passenger
  * **`Small Group`**: 2 to 4 passengers
  * **`Large Group`**: 5+ passengers
* **`passenger_category`**: Stores the readable text category (`Single`, `Small Group`, or `Large Group`).
* **`passenger_category_label`**: Converts the string categories into machine-readable integer codes using `LabelEncoder()`:
  * **`0`**: Large Group
  * **`1`**: Single
  * **`2`**: Small Group

In [18]:
# Apply One-Hot Encoding on the categorical column
df_onehot = pd.get_dummies(df, columns=['passenger_category'], prefix='passenger', dtype=int)

# Preview the newly created binary dummy columns
df_onehot.filter(like='passenger').head()

,passenger_count,passenger_category_label,passenger_Large Group,passenger_Single,passenger_Small Group
0,1,1,0,1,0
1,1,1,0,1,0
2,1,1,0,1,0
3,1,1,0,1,0
4,1,1,0,1,0


### Feature Encoding 

* **`passenger_count` (`1`)**: The original raw feature value indicating a single passenger.
* **`passenger_category_label` (`1`)**: Result of **Label Encoding**, where categories are mapped to integers based on alphabetical order (`0`: Large Group, `1`: Single, `2`: Small Group).
* **`passenger_Large Group` (`0`)**: **One-Hot Encoding** flag — `0` means False (not a large group).
* **`passenger_Single` (`1`)**: **One-Hot Encoding** flag — `1` means True (is a single passenger).
* **`passenger_Small Group` (`0`)**: **One-Hot Encoding** flag — `0` means False (not a small group).


---
# 6. NUMPY IMPLEMENTATION 
- Convert dataset into NumPy arrays 
- Perform slicing and indexing 
- Apply aggregation functions 
- Perform mathematical operations 

In [19]:

# Convert trip_distance and trip_duration columns into a 2D NumPy array

numpy_array = df[['trip_distance', 'trip_duration']].to_numpy()

# Check shape and array type
print("Array Shape:", numpy_array.shape)
print("Data Type:", numpy_array.dtype)
print("\nFirst 5 rows:\n", numpy_array[:5])

Array Shape: (1457586, 2)
Data Type: float64

First 5 rows:
 [[ 1.49852078  7.58333333]
 [ 1.80550717 11.05      ]
 [ 6.3850985  35.4       ]
 [ 1.48549842  7.15      ]
 [ 1.18858846  7.25      ]]



---

### ``We`` Convert ['trip_distance', 'trip_duration'] to numpy array
- dtype --->  float
- array shape ----> 1457586 Rows and 2 Column

In [20]:
#Perform slicing and indexing

# 1. Slicing
print("First 5 rows:\n", numpy_array[:5])


# 2. Indexing
print("First row:\n", numpy_array[0])
print("First column:\n", numpy_array[:, 0])

First 5 rows:
 [[ 1.49852078  7.58333333]
 [ 1.80550717 11.05      ]
 [ 6.3850985  35.4       ]
 [ 1.48549842  7.15      ]
 [ 1.18858846  7.25      ]]
First row:
 [1.49852078 7.58333333]
First column:
 [1.49852078 1.80550717 6.3850985  ... 7.82460578 1.09256389 1.13404196]


In [21]:
# Apply aggregation functions on the NumPy array

# 1. Mean
mean_distance = numpy_array[:, 0].mean()
mean_duration = numpy_array[:, 1].mean()

# 2. Median
median_distance = np.median(numpy_array[:, 0])
median_duration = np.median(numpy_array[:, 1])

# 3. Standard Deviation
std_distance = numpy_array[:, 0].std()
std_duration = numpy_array[:, 1].std()

#Max, Min, and Sum directly using NumPy.

max_distance = numpy_array[:, 0].max()
min_distance = numpy_array[:, 0].min()
sum_distance = numpy_array[:, 0].sum()

max_duration = numpy_array[:, 1].max()
min_duration = numpy_array[:, 1].min()
sum_duration = numpy_array[:, 1].sum()



In [22]:


print("Mean Distance:", mean_distance)
print("Mean Duration:", mean_duration)
print("\n")

print("Median Distance:", median_distance)
print("Median Duration:", median_duration)
print("\n")

print("Standard Deviation Distance:", std_distance)
print("Standard Deviation Duration:", std_duration)
print("\n")

print("Max Distance:", max_distance)
print("Min Distance:", min_distance)
print("Sum Distance:", sum_distance)
print("\n")

print("Max Duration:", max_duration)
print("Min Duration:", min_duration)
print("Sum Duration:", sum_duration)


Mean Distance: 3.4231840651843823
Mean Duration: 15.975417013701653


Median Distance: 2.092887446209374
Median Duration: 11.033333333333333


Standard Deviation Distance: 3.884148767370183
Standard Deviation Duration: 87.29254347653627


Max Distance: 47.765128865632086
Min Distance: 0.0
Sum Distance: 4989585.168835843


Max Duration: 58771.36666666667
Min Duration: 0.016666666666666666
Sum Duration: 23285544.183333337


In [23]:
# Perform mathematical operations on the NumPy array

# 1. Add
add_result = numpy_array[:, 0] + numpy_array[:, 1]

# 2. Subtract
subtract_result = numpy_array[:, 0] - numpy_array[:, 1]

# 3. Multiply
multiply_result = numpy_array[:, 0] * numpy_array[:, 1]

# 4. Divide
divide_result = numpy_array[:, 0] / numpy_array[:, 1]

print("Addition Result:\n", add_result)
print("\nSubtraction Result:\n", subtract_result)
print("\nMultiplication Result:\n", multiply_result)
print("\nDivision Result:\n", divide_result)

Addition Result:
 [ 9.08185411 12.85550717 41.7850985  ... 20.55793911  7.30923056
  4.43404196]

Subtraction Result:
 [ -6.08481255  -9.24449283 -29.0149015  ...  -4.90872756  -5.12410277
  -2.16595804]

Multiplication Result:
 [ 11.36378258  19.95085422 226.03248673 ...  99.63331354   6.79210553
   3.74233847]

Division Result:
 [0.19760714 0.16339431 0.18037001 ... 0.61449784 0.17574754 0.34364908]


In [24]:
df.info()


<class 'pandas.DataFrame'>
Index: 1457586 entries, 0 to 1458643
Data columns (total 17 columns):
 #   Column                    Non-Null Count    Dtype         
---  ------                    --------------    -----         
 0   id                        1457586 non-null  str           
 1   vendor_id                 1457586 non-null  int64         
 2   pickup_datetime           1457586 non-null  datetime64[us]
 3   dropoff_datetime          1457586 non-null  datetime64[us]
 4   passenger_count           1457586 non-null  int64         
 5   pickup_longitude          1457586 non-null  float64       
 6   pickup_latitude           1457586 non-null  float64       
 7   dropoff_longitude         1457586 non-null  float64       
 8   dropoff_latitude          1457586 non-null  float64       
 9   store_and_fwd_flag        1457586 non-null  str           
 10  trip_duration             1457586 non-null  float64       
 11  hour                      1457586 non-null  int32         
 12  da


---
# 7. ADVANCED VISUALIZATION (PLOTLY) 
- Create Geo map of pickup locations 
- Create animated plot (trips over time) 
- Create Sunburst plot (Passenger → Time → Distance category) 
- Save at least one plot as GIF 


In [25]:
sample = df.sample(5000, random_state=42)

fig = px.scatter_map(
    sample,
    lat='pickup_latitude',
    lon='pickup_longitude',
    color='passenger_count',
    zoom=10,
    title='NYC Taxi Pickup Locations'
)

fig.show()

NameError: name 'px' is not defined

In [ ]:
trips_by_hour_day = (
    df.groupby(["day_of_week", "hour"])
    .size()
    .reset_index(name="trip_count")
)

fig2 = px.bar(
    trips_by_hour_day,
    x="hour",
    y="trip_count",
    color="hour",
    animation_frame="day_of_week",
    range_y=[0, trips_by_hour_day["trip_count"].max() * 1.1],
    title="Trips by Hour, Animated Across Days of Week"
)
fig2.show()

### Interpretation: Trips by Hour, Animated Across Days of Week

- This chart shows how many taxi trips happen at each hour of the day, and you can click play to see how it changes day by day.
 
- **Very few trips happen late at night** (around 1 AM to 5 AM) — most people are asleep.
  
- **Trips start increasing in the morning**, especially after 6–7 AM when people start their day.
  
- **The busiest time is in the evening** (around 5 PM to 9 PM) — likely people going home from work or going out.
  
- This pattern looks **almost the same every day** of the week — quiet at night, busy in the evening.
  
- **Takeaway:** People take the most taxis in the evening and the fewest late at night.

In [ ]:
def time_bucket(h):
    if 5 <= h < 12: return "Morning"
    elif 12 <= h < 17: return "Afternoon"
    elif 17 <= h < 21: return "Evening"
    else: return "Night"

df["time_of_day"] = df["hour"].apply(time_bucket)

df["distance_category"] = pd.cut(
    df["trip_distance"],
    bins=[-0.01, 2, 6, float("inf")],   # -0.01 so 0-distance trips are included
    labels=["Short", "Medium", "Long"]
)

# Safety net: drop any remaining NaNs (e.g. negative distances, bad data)
df_sunburst = df.dropna(subset=["passenger_category_label", "time_of_day", "distance_category"])

fig3 = px.sunburst(
    df_sunburst,
    path=["passenger_category_label", "time_of_day", "distance_category"],
    title="Passenger → Time of Day → Distance Category"
)
fig3.show()


 
 


### Interpretation: Passenger → Time of Day → Distance Category

- The sunburst shows how trips break down by **passenger category** (inner ring), then **time of day** (middle ring), then **trip distance** (outer ring).
- Ring size = number of trips — bigger slice means more trips fall in that group.
- **Inner ring:** **Single passengers (1)** make up the largest share of trips, followed by **small groups (0)**, while **large groups (2)** account for the smallest share.
- **Middle ring:** within each passenger category, trips are fairly spread across Morning, Afternoon, Evening, and Night, with no single time slot dominating heavily.
- **Outer ring:** **Short-distance trips are the most common** across almost every passenger–time combination, followed by Medium, with Long trips making up the smallest share.
- **Takeaway:** Most taxi rides are taken by single passengers on short trips, regardless of the time of day.